identity/app/controllers/AccountDeletionController.scala

```diff
@@ -0,0 +1,162 @@
+        emailPasswdAuth: EmailPassword,
+        idRequest: IdentityRequest)(implicit request: AuthenticatedActions.AuthRequest[A]) = {
+
+      sealed trait AccountDeletionFailures
+      case object FailedToEnterCorrectPassword extends AccountDeletionFailures
+      case object FailedToUnauthenticateUser extends AccountDeletionFailures
+      case object FailedToDeleteAccount extends AccountDeletionFailures
+      case object FailedToRemoveFromAllEmailLists extends AccountDeletionFailures
+
+      def checkUserEnteredCorrectPassword(): EitherT[Future, AccountDeletionFailures, Unit] =
+        (for {
+          auth <- EitherT.fromEither(identityApiClient.authBrowser(emailPasswdAuth, idRequest.trackingData))
+          _ <- EitherT.fromEither(signinService.getCookies(Future.successful(Right(auth)), true))
+        } yield ()).leftMap(_ => FailedToEnterCorrectPassword)
+
+      def unauthenticateUser(): EitherT[Future, AccountDeletionFailures, CookiesResponse] =
+        EitherT.fromEither(api.unauth(emailPasswdAuth, idRequest.trackingData)).leftMap(_ => FailedToUnauthenticateUser)
+
+      def removeFromAllMalingLists(): EitherT[Future, AccountDeletionFailures, ETSubscriber] =
+        EitherT(exactTargetService.unsubscribeFromAllLists(emailPasswdAuth.email)).leftMap(_ => FailedToRemoveFromAllEmailLists)
+
+      def deleteAccountProper(): EitherT[Future, AccountDeletionFailures, Unit] =
+        EitherT.fromEither(identityApiClient.deleteAccount(request.user.auth, emailPasswdAuth)).leftMap(_ => FailedToDeleteAccount)
+
+      def clearCookiesAndDisplaySuccessForm() =
+        NoCache(SeeOther(routes.AccountDeletionController.renderAccountDeletionConfirmForm().url))
+          .discardingCookies(cookiesToDiscard: _*)
+
+      def processFailures(failure: AccountDeletionFailures) = failure match {
+          case FailedToEnterCorrectPassword =>
+            SeeOther(routes.AccountDeletionController.renderAccountDeletionForm ().url)
+              .flashing(boundForm.withError ("password", "Password is incorrect").toFlash)
+
+          case FailedToUnauthenticateUser =>
+            logger.error(s"Failed to un-authenticate user ${request.user.user.id} during account deletion.")
+            renderFormWithUnableToDeleteAccountError(boundForm)
+
+          case FailedToRemoveFromAllEmailLists =>
+            logger.error(s"Failed to remove user ${request.user.user.id} from all mailing lists during account deletion.")
+            renderFormWithUnableToDeleteAccountError(boundForm)
+
+          case FailedToDeleteAccount =>
+            logger.error(s"Failed to delete account for user ${request.user.user.id}")
+            renderFormWithUnableToDeleteAccountError(boundForm)
+        }
+
+      (for {
+        _ <- checkUserEnteredCorrectPassword()
+        _ <- removeFromAllMalingLists()
+        _ <- unauthenticateUser()
+        _ <- deleteAccountProper()
+      } yield ()).fold(processFailures, _ => clearCookiesAndDisplaySuccessForm)
```

Comment:
If you put the same type in Left and Right (e.g. by removing the processFailures indirection and returning the error result directly at each step) you'll have a `MergableEither` that you can just call `.merge` on instead of needing to fold.